File: api/app/schemas/modificacion.py
from pydantic import BaseModel, ConfigDict, field_validator
from uuid import UUID
from datetime import date, datetime
from decimal import Decimal

TIPOS_MODIFICACION = ["Habilitacion", "Anulacion", "Credito Suplementario", "Transferencia"]
ESTADOS_MODIFICACION = ["aprobada", "pendiente", "anulada"]


class ModificacionBase(BaseModel):
    entidad_id: UUID
    numero_resolucion: str
    tipo_modificacion: str
    fecha_aprobacion: date
    monto_total: Decimal
    descripcion: str | None = None
    estado: str = "aprobada"

    @field_validator("tipo_modificacion")
    @classmethod
    def validar_tipo(cls, v: str) -> str:
        if v not in TIPOS_MODIFICACION:
            raise ValueError(f"Tipo debe ser uno de: {', '.join(TIPOS_MODIFICACION)}")
        return v

    @field_validator("estado")
    @classmethod
    def validar_estado(cls, v: str) -> str:
        if v not in ESTADOS_MODIFICACION:
            raise ValueError(f"Estado debe ser uno de: {', '.join(ESTADOS_MODIFICACION)}")
        return v

    @field_validator("numero_resolucion")
    @classmethod
    def validar_resolucion(cls, v: str) -> str:
        v = v.strip()
        if len(v) < 3:
            raise ValueError("El número de resolución es muy corto")
        return v

    @field_validator("monto_total")
    @classmethod
    def validar_monto(cls, v: Decimal) -> Decimal:
        if v < 0:
            raise ValueError("El monto debe ser mayor o igual a 0")
        return v


class ModificacionCreate(ModificacionBase):
    pass


class ModificacionUpdate(BaseModel):
    numero_resolucion: str | None = None
    tipo_modificacion: str | None = None
    fecha_aprobacion: date | None = None
    monto_total: Decimal | None = None
    descripcion: str | None = None
    estado: str | None = None


class ModificacionOut(ModificacionBase):
    id: UUID
    creado_en: datetime
    actualizado_en: datetime | None = None
    entidad_nombre: str | None = None

    model_config = ConfigDict(from_attributes=True)

File: web/src/types/modificacion.ts
import { z } from 'zod';

export const TIPOS_MODIFICACION = [
  'Habilitacion',
  'Anulacion',
  'Credito Suplementario',
  'Transferencia',
] as const;

export const ESTADOS_MODIFICACION = ['aprobada', 'pendiente', 'anulada'] as const;

export type TipoModificacion = (typeof TIPOS_MODIFICACION)[number];
export type EstadoModificacion = (typeof ESTADOS_MODIFICACION)[number];

export const modificacionSchema = z.object({
  entidad_id: z.string().uuid('Debe seleccionar una entidad'),
  numero_resolucion: z
    .string()
    .min(3, 'Mínimo 3 caracteres')
    .max(50, 'Máximo 50 caracteres'),
  tipo_modificacion: z.string().min(1, 'Seleccione un tipo válido').refine((val) => TIPOS_MODIFICACION.includes(val as TipoModificacion), {
    message: 'Seleccione un tipo válido',
  }),
  fecha_aprobacion: z.string().min(1, 'La fecha es requerida'),
  monto_total: z.coerce
    .number()
    .min(0, 'Debe ser mayor o igual a 0'),
  descripcion: z.string().optional().or(z.literal('')),
  estado: z.enum(ESTADOS_MODIFICACION).default('aprobada'),
});

export type ModificacionFormValues = z.infer<typeof modificacionSchema>;

export interface ModificacionEntity {
  id: string;
  entidad_id: string;
  numero_resolucion: string;
  tipo_modificacion: TipoModificacion;
  fecha_aprobacion: string;
  monto_total: number;
  descripcion?: string;
  estado: EstadoModificacion;
  creado_en: string;
  actualizado_en?: string;
  entidad_nombre?: string;
}

export interface ModificacionPaginatedResponse {
  data: ModificacionEntity[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
}

export interface ModificacionListParams {
  page: number;
  limit: number;
  search?: string;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
  entidad_id?: string;
  tipo_modificacion?: string;
  estado?: string;
  anio?: number;
}